File: logcal/Services/CloudSyncService.swift
//
//  CloudSyncService.swift
//  logcal
//
//  Created by Antriksh Johri on 15/12/25.
//

import Foundation
import SwiftData
import FirebaseAuth
import Combine

@MainActor
class CloudSyncService: ObservableObject {
    private let firestoreService = FirestoreService()
    @Published var isSyncing: Bool = false
    @Published var syncError: String?
    @Published var lastSyncTime: Date = Date() // Trigger view updates when sync completes
    private var currentUserId: String? // Track current authenticated user ID
    private var isAnonymousSession: Bool = false // Track if we're in anonymous mode
    private let userDefaults = UserDefaults.standard
    private let lastUserIdKey = "lastSyncedUserId" // Persist user ID to detect account switches
    
    /// Sync meal entries to Firestore (save new ones)
    func syncMealToCloud(_ entry: MealEntry) async {
        // Only sync if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping cloud sync")
            return
        }
        
        do {
            try await firestoreService.saveMealEntry(entry)
            print("DEBUG: Successfully synced meal to cloud: \(entry.id)")
        } catch {
            print("DEBUG: Error syncing meal to cloud: \(error)")
            syncError = "Failed to sync to cloud: \(error.localizedDescription)"
        }
    }
    
    /// Fetch all meals from Firestore and merge with local data
    func syncFromCloud(modelContext: ModelContext) async {
        // Only sync if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping cloud sync")
            // Clear authenticated session state
            currentUserId = nil
            return
        }
        
        let newUserId = user.uid
        
        // Check if we're switching from anonymous to authenticated
        if isAnonymousSession {
            print("DEBUG: Switching from anonymous to authenticated user, keeping local guest data to merge...")
            isAnonymousSession = false
        }
        
        // Get last synced user ID from UserDefaults (persists across sign outs)
        let lastSyncedUserId = userDefaults.string(forKey: lastUserIdKey)
        
        // Check if user has changed (different authenticated account signed in)
        if let previousUserId = currentUserId, previousUserId != newUserId {
            print("DEBUG: User changed from \(previousUserId) to \(newUserId), clearing local data first...")
            await clearLocalMeals(modelContext: modelContext)
        }
        // If currentUserId is nil but we have a different lastSyncedUserId, user switched accounts
        else if currentUserId == nil, let lastUserId = lastSyncedUserId, lastUserId != newUserId {
            print("DEBUG: Different user signing in (last: \(lastUserId), new: \(newUserId)), clearing local data first...")
            await clearLocalMeals(modelContext: modelContext)
        }
        // If currentUserId is nil and no lastSyncedUserId, but we have local data, clear it (fresh sign in with old data)
        else if currentUserId == nil, lastSyncedUserId == nil {
            let descriptor = FetchDescriptor<MealEntry>()
            if let localMeals = try? modelContext.fetch(descriptor), !localMeals.isEmpty {
                print("DEBUG: New user signing in with existing local data, clearing local data first...")
                await clearLocalMeals(modelContext: modelContext)
            }
        }
        
        // Update current user ID and session type
        currentUserId = newUserId
        isAnonymousSession = false
        
        // Persist user ID to UserDefaults to detect account switches after sign out
        userDefaults.set(newUserId, forKey: lastUserIdKey)
        
        isSyncing = true
        syncError = nil
        
        do {
            // #region agent log
            DebugLogger.log(location: "CloudSyncService.swift:80", message: "Starting fetch from Firestore", data: ["userId": newUserId], hypothesisId: "B")
            // #endregion
            // Fetch from Firestore
            let cloudMeals = try await firestoreService.fetchMealEntries()
            // #region agent log
            DebugLogger.log(location: "CloudSyncService.swift:82", message: "Fetched meals from Firestore", data: ["mealCount": cloudMeals.count], hypothesisId: "B")
            // #endregion
            print("DEBUG: Fetched \(cloudMeals.count) meals from cloud")
            
            // Get local meals
            let descriptor = FetchDescriptor<MealEntry>()
            let localMeals = try modelContext.fetch(descriptor)
            print("DEBUG: Found \(localMeals.count) local meals")
            
            // Create a set of local meal IDs for quick lookup
            let localMealIds = Set(localMeals.map { $0.id })
            
            // Add cloud meals that don't exist locally
            var addedCount = 0
            for cloudMeal in cloudMeals {
                if !localMealIds.contains(cloudMeal.id) {
                    modelContext.insert(cloudMeal)
                    addedCount += 1
                }
            }
            
            if addedCount > 0 {
                // #region agent log
                DebugLogger.log(location: "CloudSyncService.swift:102", message: "Saving meals to modelContext", data: ["addedCount": addedCount], hypothesisId: "B")
                // #endregion
                // Save the context
                try modelContext.save()
                // #region agent log
                DebugLogger.log(location: "CloudSyncService.swift:104", message: "Saved meals to modelContext", data: ["addedCount": addedCount], hypothesisId: "B")
                // #endregion
                print("DEBUG: Added \(addedCount) meals from cloud to local storage")
                
                // Verify the save worked by checking local meals again
                // Use a fresh fetch to ensure we're reading from the persisted store
                let verifyDescriptor = FetchDescriptor<MealEntry>()
                let verifyMeals = try modelContext.fetch(verifyDescriptor)
                print("DEBUG: After save, found \(verifyMeals.count) local meals")
                
                // #region agent log
                DebugLogger.log(location: "CloudSyncService.swift:122", message: "Verification after save", data: ["expectedCount": addedCount, "actualCount": verifyMeals.count], hypothesisId: "B")
                // #endregion
                
                // If verification shows 0 meals, there's a modelContext issue
                if verifyMeals.count == 0 && addedCount > 0 {
                    print("ERROR: modelContext.save() did not persist meals! This indicates a modelContext issue.")
                    // Try saving again
                    try modelContext.save()
                    let retryMeals = try modelContext.fetch(verifyDescriptor)
                    print("DEBUG: After retry save, found \(retryMeals.count) local meals")
                }
                
                // Small delay to ensure SwiftData propagates changes to @Query
                try? await Task.sleep(nanoseconds: 200_000_000) // 0.2 seconds
                
                // Update lastSyncTime to trigger onChange handlers in HistoryView
                // This will cause the view to check if refresh is needed
                lastSyncTime = Date()
                
                // Another small delay to ensure the view refresh happens
                try? await Task.sleep(nanoseconds: 100_000_000) // 0.1 seconds
            } else {
                print("DEBUG: No new meals to add from cloud")
            }
            
            isSyncing = false
            // Update lastSyncTime even if no meals were added to trigger any pending updates
            if addedCount == 0 {
                lastSyncTime = Date()
            }
        } catch {
            print("DEBUG: Error syncing from cloud: \(error)")
            syncError = "Failed to sync from cloud: \(error.localizedDescription)"
            isSyncing = false
        }
    }
    
    /// Migrate all local meals to cloud (for when user first signs in)
    func migrateLocalToCloud(modelContext: ModelContext) async {
        // Only migrate if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping migration")
            return
        }
        
        isSyncing = true
        syncError = nil
        
        do {
            // Get all local meals
            let descriptor = FetchDescriptor<MealEntry>()
            let localMeals = try modelContext.fetch(descriptor)
            print("DEBUG: Migrating \(localMeals.count) local meals to cloud")
            
            if !localMeals.isEmpty {
                try await firestoreService.syncLocalMealsToCloud(entries: localMeals)
                print("DEBUG: Successfully migrated \(localMeals.count) meals to cloud")
            }
            
            // Also migrate daily goal if it exists in UserDefaults
            // Note: We can't access @AppStorage here, so we'll read from UserDefaults directly
            let localDailyGoal = UserDefaults.standard.double(forKey: "dailyGoal")
            if localDailyGoal > 0 && localDailyGoal != 2000 { // Only migrate if it's been changed from default
                print("DEBUG: Migrating daily goal to cloud: \(localDailyGoal)")
                try await firestoreService.saveDailyGoal(localDailyGoal)
                print("DEBUG: Successfully migrated daily goal to cloud")
            }
            
            isSyncing = false
        } catch {
            print("DEBUG: Error migrating to cloud: \(error)")
            syncError = "Failed to migrate to cloud: \(error.localizedDescription)"
            isSyncing = false
        }
    }
    
    /// Delete meal from cloud
    func deleteMealFromCloud(_ entry: MealEntry) async {
        // Only delete from cloud if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping cloud delete")
            return
        }
        
        do {
            try await firestoreService.deleteMealEntry(entry)
            print("DEBUG: Successfully deleted meal from cloud: \(entry.id)")
        } catch {
            print("DEBUG: Error deleting meal from cloud: \(error)")
            syncError = "Failed to delete from cloud: \(error.localizedDescription)"
        }
    }
    
    /// Clear all local meals (used when user signs out or switches accounts)
    func clearLocalMeals(modelContext: ModelContext) async {
        print("DEBUG: Clearing all local meals...")
        
        let container = modelContext.container
        
        // CRITICAL: Perform deletion on a BACKGROUND THREAD using Task.detached.
        //
        // The root cause of the EXC_BREAKPOINT crash:
        // 1. We delete MealEntry objects and call save()
        // 2. save() internally posts an NSNotification via NSNotificationCenter
        // 3. NSNotificationCenter delivers notifications SYNCHRONOUSLY on the posting thread
        // 4. SwiftData_SwiftUI's @Query observer receives the notification on the main thread
        // 5. The observer tries to diff/process the deleted objects and crashes
        //
        // By using Task.detached with a separate ModelContext:
        // - The ModelContext is created and used entirely on a background thread
        // - save() posts the notification on the background thread
        // - SwiftData_SwiftUI's @Query observer (registered on main thread) processes
        //   the store change asynchronously, after the deletion is complete
        // - The @Query safely re-fetches from the (now empty) persistent store
        do {
            try await Task.detached(priority: .userInitiated) {
                let bgContext = ModelContext(container)
                bgContext.autosaveEnabled = false
                
                let descriptor = FetchDescriptor<MealEntry>()
                let localMeals = try bgContext.fetch(descriptor)
                guard !localMeals.isEmpty else {
                    print("DEBUG: No local meals to delete")
                    return
                }
                print("DEBUG: Found \(localMeals.count) local meals to delete (background thread)")
                for meal in localMeals {
                    bgContext.delete(meal)
                }
                try bgContext.save()
                print("DEBUG: Successfully cleared \(localMeals.count) local meals via background thread")
            }.value
        } catch {
            print("DEBUG: Error clearing local meals: \(error)")
        }
        
        // Allow time for the store change notification to propagate to the UI context's @Query
        try? await Task.sleep(nanoseconds: 500_000_000) // 0.5 seconds
        
        // Clear session state when clearing local data
        currentUserId = nil
        isAnonymousSession = false
        
        // Clear persisted user ID when clearing data
        userDefaults.removeObject(forKey: lastUserIdKey)
    }
    
    /// Initialize anonymous session (clear authenticated data when switching to anonymous)
    func initializeAnonymousSession(modelContext: ModelContext) async {
        print("DEBUG: Initializing anonymous session...")
        
        // Note: We no longer clear local meals here. They are safely cleared in logcalApp.swift
        // when the user signs out, while the AppRootView (and its @Query observers) are 
        // completely detached from the view hierarchy. This prevents SwiftData crashes.
        
        // Set anonymous session state
        currentUserId = nil
        isAnonymousSession = true
        print("DEBUG: Anonymous session initialized")
    }
    
    /// Sync daily goal to Firestore
    func syncDailyGoalToCloud(_ goal: Double) async {
        // Only sync if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping cloud sync for daily goal")
            return
        }
        
        do {
            try await firestoreService.saveDailyGoal(goal)
            print("DEBUG: Successfully synced daily goal to cloud: \(goal)")
        } catch {
            print("DEBUG: Error syncing daily goal to cloud: \(error)")
            syncError = "Failed to sync daily goal to cloud: \(error.localizedDescription)"
        }
    }
    
    /// Fetch daily goal from Firestore
    func fetchDailyGoalFromCloud() async -> Double? {
        // Only fetch if user is signed in (not anonymous)
        guard let user = Auth.auth().currentUser, !user.isAnonymous else {
            print("DEBUG: User is anonymous or not signed in, skipping cloud fetch for daily goal")
            return nil
        }
        
        print("DEBUG: Fetching daily goal from cloud for user: \(user.uid)")
        do {
            let goal = try await firestoreService.fetchDailyGoal()
            print("DEBUG: fetchDailyGoalFromCloud result: \(goal?.description ?? "nil")")
            return goal
        } catch {
            print("DEBUG: Error fetching daily goal from cloud: \(error)")
            syncError = "Failed to fetch daily goal from cloud: \(error.localizedDescription)"
            return nil
        }
    }
}

File: android/app/src/main/java/com/serene/logcal/service/CloudSyncService.kt
package com.serene.logcal.service

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import com.serene.logcal.data.local.MealEntryEntity
import com.serene.logcal.data.local.SavedMealEntity
import com.serene.logcal.data.repository.AppGraph
import com.serene.logcal.util.DebugLogger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

class CloudSyncService(private val context: Context) {
    private val firestoreService = FirestoreService()
    private val auth = FirebaseAuth.getInstance()
    private val localRepo = AppGraph.localMealRepository(context)
    private val favRepo = AppGraph.localSavedMealsRepository(context)
    private val prefManager = AppGraph.preferenceManager(context)
    private val json = Json { ignoreUnknownKeys = true }

    private val _isSyncing = MutableStateFlow(false)
    val isSyncing: StateFlow<Boolean> = _isSyncing.asStateFlow()

    private val _syncError = MutableStateFlow<String?>(null)
    val syncError: StateFlow<String?> = _syncError.asStateFlow()

    private val _lastSyncTimeMillis = MutableStateFlow(0L)
    val lastSyncTimeMillis: StateFlow<Long> = _lastSyncTimeMillis.asStateFlow()

    suspend fun syncMealToCloud(entry: MealEntryEntity) {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) {
            DebugLogger.d("DEBUG: [CloudSyncService] User is anonymous or not signed in, skipping syncMealToCloud")
            return
        }
        try {
            firestoreService.saveMealEntry(entry)
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error syncing meal to cloud", e)
            _syncError.value = "Failed to sync meal: ${e.localizedMessage}"
        }
    }

    suspend fun syncFromCloud() {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) {
            DebugLogger.d("DEBUG: [CloudSyncService] User is anonymous or not signed in, skipping syncFromCloud")
            return
        }

        val newUserId = user.uid
        val lastSyncedUserId = prefManager.lastSyncedUserId

        // Clear local meals if account changed
        if (lastSyncedUserId != null && lastSyncedUserId != newUserId) {
            DebugLogger.d("DEBUG: [CloudSyncService] User changed, clearing local data...")
            clearLocalMeals()
        }

        prefManager.lastSyncedUserId = newUserId

        // Guard against duplicate concurrent syncs
        if (_isSyncing.value) {
            DebugLogger.d("DEBUG: [CloudSyncService] Already syncing, skipping duplicate sync request.")
            return
        }

        _isSyncing.value = true
        _syncError.value = null

        try {
            // 1. Sync meals
            val cloudMeals = firestoreService.fetchMealEntries()
            val localMeals = localRepo.observeHistoryMeals().first()
            val localMealIds = localMeals.map { it.id }.toSet()

            val toAdd = mutableListOf<MealEntryEntity>()
            for (cloudMeal in cloudMeals) {
                if (cloudMeal.id !in localMealIds) {
                    toAdd.add(cloudMeal)
                }
            }

            if (toAdd.isNotEmpty()) {
                localRepo.insertMeals(toAdd)
                DebugLogger.d("DEBUG: [CloudSyncService] Merged ${toAdd.size} meals from cloud to local Room DB")
            }

            // 2. Sync favorites
            val cloudFavs = firestoreService.fetchSavedMealsFromCloud()
            val localFavs = favRepo.getAll()
            val localFavIds = localFavs.map { it.id }.toSet()

            val favsToAdd = mutableListOf<SavedMealEntity>()
            for (cloudFav in cloudFavs) {
                if (cloudFav.id !in localFavIds) {
                    favsToAdd.add(cloudFav)
                } else {
                    val local = localFavs.find { it.id == cloudFav.id }
                    if (local != null && (local.title != cloudFav.title || local.displayOrder != cloudFav.displayOrder)) {
                        favRepo.save(cloudFav) // overwrite
                    }
                }
            }

            if (favsToAdd.isNotEmpty()) {
                favRepo.saveAll(favsToAdd)
                DebugLogger.d("DEBUG: [CloudSyncService] Merged ${favsToAdd.size} favorites from cloud to local Room DB")
            }

            _lastSyncTimeMillis.value = System.currentTimeMillis()
            _isSyncing.value = false
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error syncing from cloud", e)
            _syncError.value = "Failed to sync from cloud: ${e.localizedMessage}"
            _isSyncing.value = false
        }
    }

    suspend fun migrateLocalToCloud() {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) {
            DebugLogger.d("DEBUG: [CloudSyncService] User is anonymous, skipping migration")
            return
        }

        _isSyncing.value = true
        _syncError.value = null

        try {
            // Migrate meals
            val localMeals = localRepo.observeHistoryMeals().first()
            if (localMeals.isNotEmpty()) {
                val entities = localMeals.map { meal ->
                    MealEntryEntity(
                        id = meal.id,
                        timestampMillis = meal.timestampMillis,
                        createdAtMillis = meal.createdAtMillis,
                        foodText = meal.foodText,
                        mealType = meal.mealType,
                        totalCalories = meal.totalCalories,
                        rawResponseJson = json.encodeToString(com.serene.logcal.model.MealLogResponse.serializer(), meal.response),
                        hasImage = meal.hasImage
                    )
                }
                firestoreService.syncLocalMealsToCloud(entities)
            }

            // Migrate preferences
            val dailyGoal = prefManager.dailyGoal
            val protein = prefManager.proteinGoal
            val carbs = prefManager.carbsGoal
            val fat = prefManager.fatGoal
            val style = prefManager.dietStyle
            if (dailyGoal > 0) {
                firestoreService.saveDailyGoal(dailyGoal, protein, carbs, fat, style)
            }

            // Migrate favorites
            val localFavs = favRepo.getAll()
            if (localFavs.isNotEmpty()) {
                firestoreService.saveSavedMealsToCloud(localFavs)
            }

            _isSyncing.value = false
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error migrating local data to cloud", e)
            _syncError.value = "Failed to migrate data: ${e.localizedMessage}"
            _isSyncing.value = false
        }
    }

    suspend fun deleteMealFromCloud(id: String) {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) return
        try {
            firestoreService.deleteMealEntry(id)
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error deleting meal from cloud", e)
            _syncError.value = "Failed to delete from cloud: ${e.localizedMessage}"
        }
    }

    suspend fun clearLocalMeals() = withContext(Dispatchers.IO) {
        localRepo.deleteAllMeals()
        favRepo.deleteAll()
        prefManager.lastSyncedUserId = null
    }

    suspend fun initializeAnonymousSession() {
        prefManager.lastSyncedUserId = null
        DebugLogger.d("DEBUG: [CloudSyncService] Anonymous session initialized")
    }

    suspend fun syncDailyGoalToCloud(goal: Double) {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) return
        try {
            firestoreService.saveDailyGoal(goal)
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error syncing daily goal", e)
        }
    }

    suspend fun syncUserPreferencesToCloud(
        dailyGoal: Double,
        proteinGoal: Double,
        carbsGoal: Double,
        fatGoal: Double,
        dietStyle: String
    ) {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) return
        try {
            firestoreService.saveDailyGoal(dailyGoal, proteinGoal, carbsGoal, fatGoal, dietStyle)
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error syncing preferences", e)
        }
    }

    suspend fun syncSavedMealsToCloud() {
        val user = auth.currentUser
        if (user == null || user.isAnonymous) return
        try {
            val localFavs = favRepo.getAll()
            firestoreService.saveSavedMealsToCloud(localFavs)
        } catch (e: Exception) {
            DebugLogger.e("DEBUG: [CloudSyncService] Error syncing saved meals to cloud", e)
        }
    }
}
